InventoryMoveItemEvent improvements

Phoenix616 opened the pull request:
This patch adds the ability to configure the behaviour of the InventoryMoveItemEvent or disable the calling of it completely. It currently tries to keep the changes to the CraftBukkit behaviour to a minimum could be implemented way better when rewriting the whole item move logic.

The main issue that this patch tries to solve is the InventoryMoveItemEvent being called even 'though the destination inventory wouldn't actually be able to receive it. That means that hoppers pointing into full chests will call this event every few as long as it has items in it. This is bad for protection plugins like LWC that do owner lockups on every attempt to move items even when they wouldn't care about items not being moved.

The default is to just call the InventoryMoveItemEvent as cancelled when Vanilla wouldn't move the item, this matches the behaviour of other Bukkit events (like the PlayerInteractEvent) and gives plugins the ability to decide whether or not they want to listen on cancelled events. (Hint: LWC doesn't listen on cancelled events and I majorly improved the impact of hoppers from [60%](https://timings.aikar.co/v2/?id=012a429778ce4d13852952f67a2a5775&start=1476314971&end=1476315059&section=lag) to only [20%](https://timings.aikar.co/v2/?id=a3e7a12374f3402fa93066c791ad341a&start=1476315767&end=1476316187&section=lag) with an extremely large sorting machine on my server, keeping the server at ~20 ticks instead of dropping down to 10)

One could also disable the calling of the event for full target inventories in the config or the whole event completely.

Phoenix616 committed:
Add InventoryMoveItemEvent-improvements.patch

Phoenix616 committed:
Update comment (and e-mail address)

Phoenix616 committed:
Group settings together

kashike commented on Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch:
```diff
@@ -0,0 +1,165 @@
++        callMoveItemEvent = getBoolean("move-item-event.enabled", true);
++    }
++
++    public boolean callMoveItemEventWithoutMove = true;
```
Group the fields together, only add one method to populate them.

Phoenix616 replied:
Uh, yeah. I started out with a single setting and forgot about the grouping.

kashike commented on Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch:
```diff
@@ -0,0 +1,165 @@
++    }
++
++    // Vanilla method: c(IInventory, ItemStack, int, EnumDirection)
++    private static boolean enoughRoomInSlot(IInventory iinventory, ItemStack itemstack, int i, EnumDirection enumdirection) {
```
Where does this method come from? If the method body is just a copy from another method, just call that method from this new method.

Phoenix616 replied:
Same as with addItem, the checking code is inside the c method alongside the actual item movements. You'd have to split the full Vanilla logic into two different methods.

zachbr replied:
There is precedent to do so, I'd rather see it split than a small portion copied. Unless that split has severe implications on readability. This will also help us maintain it.

kashike commented on Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch:
```diff
@@ -0,0 +1,165 @@
+ 
++    // Paper start - method to check whether or not an inventory will accept the provided item. This should match Vanilla behaviour
++    // Vanilla method: addItem(IInventory, ItemsStack, EnumDirection)
++    static boolean willAcceptItem(IInventory iinventory, ItemStack itemstack, EnumDirection enumdirection) {
```
Where does this method come from? If the method body is just a copy from another method, just call that method from this new method.

Phoenix616 replied:
This method only pulls the code that actually checks the inventory out of the addItem method which contains also the code to actually move the item. A better way might be the rewriting of the full addItem method instead of just pulling the code out, not sure how the policy is about such situation.

zachbr replied:
Split and rewrite as needed, not too keen on duplicating vanilla logic.

Phoenix616 committed:
Merge

Phoenix616 committed:
Remove .orig. Don't trust tab complete at 3am

Phoenix616 committed:
Remove Vanilla logic duplication

AlfieC commented on Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch:
```diff
@@ -0,0 +1,165 @@
+-            ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++            // Paper start - call event as cancelled when inventory wont accept item in Vanilla
++            if (ihopper.getWorld().paperConfig.callMoveItemEvent) {
++                boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
```
You can save a little CPU time by moving the check into the if statement so it isn't calculated if `callMoveItemEventWithoutMove` is enabled.

Phoenix616 replied:
It's not as easy as that. The check needs to be also made when callMoveEventCancelled (withoutMove, maybe the variable name should be more descriptive) is true.

The code would have to be something like this:

```
boolean acceptsItem = true;
if (!world.paperConfig.callMoveItemEventWithoutMove || world.paperConfig.callMoveEventCancelledWithoutMove) {
    acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
}
if (world.paperConfig.callMoveItemEventWithoutMove || acceptsItem) {
    if (world.paperConfig.callMoveEventCancelledWithoutMove && !acceptsItem) {
        event.setCancelled(true);
    }
    this.getWorld().getServer().getPluginManager().callEvent(event);
}
```

The acceptsItem variable assignment could be compressed to `boolean acceptsItem = (!world.paperConfig.callMoveItemEventWithoutMove || world.paperConfig.callMoveEventCancelledWithoutMove && willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);` but that's kinda hard to read.

I don't see how you could get the willAcceptItem into the other statement. Not sure if it is worth to expand the calculation of how the event should be called to get rid of the many nested if statements. If so it might be best to move it into its own method?

AlfieC commented on Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch:
```diff
@@ -0,0 +1,165 @@
+-                        this.getWorld().getServer().getPluginManager().callEvent(event);
++                        // Paper start - call event as cancelled when inventory wont accept item in Vanilla
++                        if (world.paperConfig.callMoveItemEvent) {
++                            boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
```
See below - can move the `willAcceptItem` method into the if statement.

Phoenix616 committed:
Cleanup and move event calculation to own method

Phoenix616 committed:
Fix called event not being returned.

Also remove unecessary clones of the itemstack in the event
constructor as the event itself uses clones in the getItem
and setItem methods.
